system/lib/libc/musl/src/thread/__timedwait.c

```diff
@@ -75,18 +78,14 @@ int __timedwait_cp(volatile int *addr, int val,
 				// cancel execution.
 				return ECANCELED;
 			}
+			// Must wait in slices in case this thread is cancelled in between.
+			r = -emscripten_futex_wait((void*)addr, val,
+				msecsToSleep > max_ms_to_sleep ? max_ms_to_sleep : msecsToSleep);
 			// Assist other threads by executing proxied operations that are effectively singlethreaded.
 			if (is_runtime_thread) emscripten_main_thread_process_queued_calls();
-			// Must wait in slices in case this thread is cancelled in between.
-			double waitMsecs = sleepUntilTime - emscripten_get_now();
-			if (waitMsecs <= 0) {
-				r = ETIMEDOUT;
-				break;
-			}
-			if (waitMsecs > 100) waitMsecs = 100; // non-main threads can sleep in longer slices.
-			if (is_runtime_thread && waitMsecs > 1) waitMsecs = 1; // the runtime thread may need to run proxied calls, so sleep in very small slices to be responsive.
-			r = -emscripten_futex_wait((void*)addr, val, waitMsecs);
-		} while(r == ETIMEDOUT);
+
+			msecsToSleep = sleepUntilTime - emscripten_get_now();
+		} while (r == ETIMEDOUT && msecsToSleep > 0);
```

Comment:
This change has the subtle difference that before, if one had a wait time of zero, then the wait loop would process `emscripten_main_thread_process_queued_calls()` and exit, but after this change, there is now an unconditional call out to `emscripten_futex_wait()`. It is uncertain what that will do, it could yield back the remaining time slice at minimum, or it could do a full kernel wait that gets triggered via the general timeout mechanism, so performance could theoretically be affected.